core/src/main/scala/org/apache/spark/sql/execution/row/RowFormatRelation.scala

```diff
@@ -82,32 +83,19 @@ class RowFormatRelation(
   @transient override lazy val region: LocalRegion =
     Misc.getRegionForTable(resolvedName, true).asInstanceOf[LocalRegion]
 
+  @transient lazy val clusterMode = SnappyContext.getClusterMode(_context.sparkContext)
   private[this] def indexedColumns: mutable.HashSet[String] = {
     val cols = new mutable.HashSet[String]()
-    val container = region.getUserAttribute.asInstanceOf[GemFireContainer]
-    val td = container.getTableDescriptor
-    if (td ne null) {
-      val baseColumns = td.getColumnNamesArray
-      val im = container.getIndexManager
-      if ((im ne null) && (im.getIndexConglomerateDescriptors ne null)) {
-        val itr = im.getIndexConglomerateDescriptors.iterator()
-        while (itr.hasNext) {
-          // first column of index has to be present in filter to be usable
-          val indexCols = itr.next().getIndexDescriptor.baseColumnPositions()
-          cols += baseColumns(indexCols(0) - 1)
-        }
-      }
-      // also add primary key
-      val primaryKey = td.getPrimaryKey
-      if (primaryKey ne null) {
-        // first column of primary key has to be present in filter to be usable
-        val pkCols = primaryKey.getKeyColumns
-        if (pkCols.nonEmpty) {
-          cols += baseColumns(pkCols(0) - 1)
-        }
-      }
+    clusterMode match {
+      case ThinClientConnectorMode(_, _) =>
+        cols ++= relInfo.indexCols
+
+      case _ =>
+        val indexCols = new Array[String](1)
+        GfxdSystemProcedures.getIndexColumns(indexCols, region)
```

Comment:
The approach is as earlier. Just moved the code to one common method than can be used both for connector or embedded mode.